### ISSUE TITLE
Confusing step in DCG left-recursion solution

### ISSUE THREAD

**Issue by jeremy-w** [triska/the-power-of-prolog#9]
> We can do this by introducing two additional arguments that let us limit the number of rule applications to the given list's length: (https://www.metalevel.at/prolog/dcg)

As a reader, it’s not clear to me why there are *two* arguments, or what’s up with the sharing between LS1 in the left/right descents.

I suspect there is one or two intermediate steps to arrive at that solution, that you are able to leap straight past thanks to running across the problem many times before. For a newcomer, though, it leaves a mystery.

If you could elaborate on how this works within the chapter, that would be very helpful!

**Comment by triska** [comment]
Thank you for this highly appropriate suggestion! The basis for this step is found in the important concept of **list&nbsp;differences**.

This was so far only briefly mentioned, and I have now added several new paragraphs about it, please have a look:

https://www.metalevel.at/prolog/dcg#leftrecursion

What do you think, does it help?

**Comment by jeremy-w** [comment]
Yes, it does. I had to read and think on it a bit, but I could follow along. I think this specific example was especially helpful:

> We can interpret the difference of these lists as describing the list of elements that are consumed by the first recursive invocation of tree_nodes//3 in this example.

Explaining it to myself, I see the two args as, arg 1 “you have this much list to work with”, arg 2 “and you left this much unused”. Then the flow from the first recursive call to the second makes sense as parceling out the overall list.

I appreciated the further words on list differences vs difference lists. The clarifications throughout the book on what constitutes modern Prolog vs legacy Prolog have been a big help in understanding what Prolog can do *today.*

**Comment by triska** [comment]
> Explaining it to myself, I see the two args as, arg 1 “you have this much list to work with”, arg 2 “and you left this much unused”

Yes, this is applicable in this case, since it assumes the list is *given* when the rule is invoked and this is indeed the main point of the example. However, this wording would not completely do justice to the versatility of the actual relation, which also works if the list is not&nbsp;specified in any&nbsp;way. In that case, it *yields* the number of tokens that are needed for this rule to apply. I am always striving for formulations that encompass all possible usage modes, and so I have worded this in such a way that both cases are subsumed, even though one direction is emphasized much more strongly.

Thank you very much for your kind feedback and for this suggestion!

**Comment by triska** [comment]
I have created a small video to explain list differences in more detail, please have a look:

https://www.metalevel.at/prolog/videos/list_differences

It's work in progress... however, I hope you find it useful!

**Comment by triska** [comment]
Also, @jeremy-w, @neuroevolutus and everyone interested: Please consider attending the **Scryer Prolog Meetup 2025** in D&uuml;sseldorf, Germany:

**https://hsd-pbsa.de/veranstaltung/scryer-prolog-meetup-2025/**

It would be great to meet you in person!

**Comment by YPares** [comment]
Hi! I had a question related to that same section. Given `tree_nodes(X) --> ...` already amounts to `tree_nodes(X, L0, L) :- ...`, wouldn't it be simpler instead of adding two extra arguments to `tree_nodes//1` to just declare `tree_nodes` as a regular predicate with explicit "state arguments" instead of a DCG?
Because the way I understand it, these two extra "state arguments" are already the list difference we want to make explicit.
Or am I missing something?

**Comment by triska** [comment]
Hi @YPares, thank you a lot for your interest! In what sense would adding the arguments be simpler than not adding them?

Also, why stop there: Would it be even simpler to directly write WAM code, since the Prolog code amounts to WAM&nbsp;instructions?

Please note also that a DCG does not necessarily amount to any specific translation to Prolog code, a system may for example be able to implement a DCG  more efficiently with a different internal convention to represent list differences. As a simple example, maybe `tree_nodes(L, L0, X)` makes internal argument passing more efficient in a system. We can only rely on `phrase/[2,3]` as interface predicates to&nbsp;DCGs.

**Comment by triska** [comment]
Also, following the tradition established in this issue, I would like to invite you <s>both</s> *all* to the **Scryer&nbsp;Prolog&nbsp;Meetup&nbsp;2026** in Vienna!

**https://www.digitalaustria.gv.at/wissenswertes/events/scryerprologmeetup2026.html**

Please attend if at all possible, it would be great to meet you in person, and I hope to see you in October in Vienna!

**Comment by YPares** [comment]
@triska My (dated) background in Prolog had DCGs presented to me strictly as syntactic sugar, not as something that could be backed by a different internal representation and thus requiring `phrase` in all cases in order not to assume some underlying representation and breaking abstraction. That's where the confusion comes from.

...and no I wouldn't want to write WAM code directly ^^. Although on a separated topic I was wondering If you considered making the WAM implementation in scryer a separately usable crate, so other projects could reuse it. There aren't a lot of mature WAM implementations in Rust currently, and even just from an educational standpoint it could be interesting. (I actually stumbled upon scryer as I was looking for that)

My post-summer work agenda is kinda like shifting sands for now but thanks for the invitation :)
